velox/expression/CastExpr.cpp

```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
   });
 }
 
+template <
+    typename TInput,
+    typename TOutput,
+    typename = typename std::enable_if_t<std::is_floating_point_v<TInput>>>
+void applyDoubleToDecimal(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<TInput>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+    const auto& sourceValue = sourceVector->valueAt(row);
+    if (!std::isfinite(sourceValue)) {
+      castResult->setNull(row, true);
+      return;
+    }
+
+    auto toValue = sourceValue *
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
```

Comment:
If the toValue is Double::max - 0.4, It will overflow and cannot get correct result.

Reply:
Postgresql convert double to varchar and then convert to numeric type, https://github.com/postgres/postgres/blob/master/src/backend/utils/adt/numeric.c#L4515

I have a PR to cast varchar to decimal https://github.com/facebookincubator/velox/pull/5307

Reply:
Here maybe cast double to int128, you can check and use part of them
https://github.com/facebookincubator/velox/blob/main/velox/type/Conversions.h#L231